traffic_ops/traffic_ops_golang/parameter/parameters.go

```diff
@@ -0,0 +1,354 @@
+package parameter
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/apache/incubator-trafficcontrol/lib/go-log"
+	"github.com/apache/incubator-trafficcontrol/lib/go-tc"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+)
+
+//we need a type alias to define functions on
+type TOParameter tc.ParameterNullable
+
+//the refType is passed into the handlers where a copy of its type is used to decode the json.
+var refType = TOParameter(tc.ParameterNullable{})
+
+func GetRefType() *TOParameter {
+	return &refType
+}
+
+//Implementation of the Identifier, Validator interface functions
+func (parameter *TOParameter) GetID() (int, bool) {
+	if parameter.ID == nil {
+		return 0, false
+	}
+	return *parameter.ID, true
+}
+
+func (parameter *TOParameter) GetAuditName() string {
+	if parameter.Name != nil {
+		return *parameter.Name
+	}
+	if parameter.ID != nil {
+		return strconv.Itoa(*parameter.ID)
+	}
+	return "unknown"
+}
+
+func (parameter *TOParameter) GetType() string {
+	return "parameter"
+}
+
+func (parameter *TOParameter) SetID(i int) {
+	parameter.ID = &i
+}
+
+func (pl *TOParameter) Validate(db *sqlx.DB) []error {
+	errs := []error{}
```

Comment:
Do you intend to be allocating an empty array here? You can almost always save the allocation with `var errs []error`, so that in the typical case, that of no errors, it doesn't do any allocation?

It's small and mostly pointless as an optimization in this particular code, but it's a good practice to be in. Also, this annoys some of the linters.

Reply:
-1

I use `errs := []error{}` not because it's an optimization, but because I think it's better to use a single way to create variables, i.e. `:=`. I'm not about to argue everyone should, and we should reject code with `var`; but I definitely prefer it, and I'd prefer we didn't demand people use `var` and use two different ways of creating variables.

Also, if you feel strongly about the allocation, `errs := []error(nil)` works too. But I'd vote we don't require that either

Reply:
`errs := []error(nil)` would be my preferred answer. It's not meaningfully longer and it avoids an allocation with effectively no work. Still... the code as it exists is fine. It's mostly a note for future work.

Reply:
also affects json decoding -- nil comes out as null; empty slice as `[]`